Xinyue/testapp p2

xinyue-uoft opened the pull request:
# Description

See issue #21 for motivation and overview.

This PR adds a standalone TPGenerator test application and the supporting test infrastructure needed to validate TPGenerator algorithms offline, independently of the full DAQ system. The new test app follows the same pattern as the existing test_tpg_processor_app.cxx (single processor test application), but is focused on end‑to‑end verification of TPGenerator using frame format data and binary validation files.

Main changes:
	•	New TPGenerator test application (test_tpg_generator_app.cxx)
	•	Implements a CLI test app (installed as tpglibs_generator_test) that:
	•	Reads binary input frames from a .bin file using FrameReader.
	•	Adapts them to DUMMY_FRAME_STRUCT via DummyFrameAdapter so they are compatible with TPGenerator.
	•	Configures TPGenerator from a JSON config (processor_configs, channel_plane_mappings, test_config)(same format as actual config)
	•	Runs frames through TPGenerator and collects output TriggerPrimitives.
	•	(Can also turn off optionally) validates the output against a .val validation file using TPReader + TPComparator.
	•	BinaryFormatSpecification.md and example README are updated to describe the TP binary format, header, sort order, and concrete usage of the new test app and datasets.

## Type of change

- [ ] Documentation (non-breaking change that adds or improves the documentation)
- [x] New feature or enhancement (non-breaking change which adds functionality)
- [ ] Optimization (non-breaking change that improves code/performance)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (whatever its nature)

## Testing checklist

- [x] Unit tests pass (e.g. `dbt-build --unittest`)
- [ ] Minimal system quicktest passes (`pytest -s minimal_system_quick_test.py`)
- [ ] Full set of integration tests pass (`daqsystemtest_integtest_bundle.sh`)
- [ ] Python tests pass if applicable (e.g. `python -m pytest`)
- [ ] Pre-commit hooks run successfully if applicable (e.g. `pre-commit run --all-files`)

This is a completely standalone test application that does not interfere with DAQ software. Test is mainly done by unit tests directed to the test application infrastructure. Two simple test cases are developed and tested, where the test application behaves as expected: 1. an all zero input sanity check, no tp 2. a square wave in a single adc channel, with tp also localized to single channel.

## Further checks

- [x] Code is commented where needed, particularly in hard-to-understand areas
- [ ] Code style is correct (`dbt-build --lint`, and/or see https://dune-daq-sw.readthedocs.io/en/latest/packages/styleguide/)
- [ ] If applicable, new tests have been added or an issue has been opened to tackle that in the future.
  (Indicate issue here: # (issue))

xinyue-uoft committed:
Add FrameReader class for reading binary frame files and associated tests.

xinyue-uoft committed:
Add DummyFrameAdapter and TPReader classes with associated tests for test frame reading infrastructure.

xinyue-uoft committed:
Add TPComparator module for comparing TPs, and its unittest.

xinyue-uoft committed:
Implement test configuration validation. Implement tpgenerator test application initial version.
Change dummy_frame_struct structure to directly use TPGenerator default format, yields a simple and clear data flow from binaries to tpgenerator.

xinyue-uoft committed:
Fixing unittest error after previous refactor commit.

xinyue-uoft committed:
Refactor out test configuration parser functionality.

xinyue-uoft committed:
Refactor the generator test applicaiton

xinyue-uoft committed:
Add max_frames consistency validation to test application,

xinyue-uoft committed:
Added some comments.

xinyue-uoft committed:
Update the BinaryFormatSpecification file to contain information on both the TPG Processor test application and the TPGenerator test application.

xinyue-uoft committed:
Improving documentation.

xinyue-uoft commented:
Eventually plan is to add the binary files and config for the two sanity type tests to DUNE asset system. To allow reviewers to perform independent tests for this PR, I will post them here together with their generation script soon.

xinyue-uoft committed:
(Could be temporary) include test generation scripts

xinyue-uoft commented:
I have temporarily included the two test binary + config + validation file generation script for the following three test cases. We can remove them after review if they shouldn't go into `tpglibs`.

### Test Patterns and Expected TP Presence

This summarizes which frames/channels are expected to have TPs for the sanity, multiframe, and golden test.

---

#### 1. Sanity test (`tpg_generator_sanity_*`)

**Input**: All ADC values are below threshold.

| Frame | Channels with TP | Notes |
|-------|------------------|-------|
| 0     | none             | No threshold crossings |

---

#### 2. Multi-frame test (`tpg_generator_multiframe_*`)

| Frame | Channels with TP | Notes |
|-------|------------------|-------|
| 0     | ch 0             | Single spike region in channel 0 |
| 1     | none             | All zeros |
| 2     | ch 16            | Single spike region in channel 16 |
| 3     | none             | All zeros |
| 4     | ch 0, ch 16, ch 32, ch 48 | Four independent spike regions |

Each spike region is contiguous within one channel, so each listed channel generates exactly one TP.

---

#### 3. Golden test (`tpg_generator_golden_*`)

| Frame | Channels with TP | Notes |
|-------|------------------|-------|
| 0     | ch 0             | Single spike region in channel 0 |
| 1     | none             | All zeros |

---

aeoranday commented on docs/testapp/BinaryFormatSpecification.md:
```diff
@@ -1,151 +1,273 @@
 
-## 3. Test File Examples
+**Key Fields in TriggerPrimitive** (for reference):
+- `time_start` (int64_t): Timestamp when TP starts
```
```suggestion
- `time_start` (uint64_t): Timestamp when TP starts
```

aeoranday commented on docs/testapp/BinaryFormatSpecification.md:
```diff
@@ -1,151 +1,273 @@
-## 3. Test File Examples
+**Key Fields in TriggerPrimitive** (for reference):
+- `time_start` (int64_t): Timestamp when TP starts
+- `channel` (channel_t, typically uint32_t): Channel number
```
The TP `channel` is 24 bits. We usually use `channel_t` (which is `uint32_t`) since it would provide the coverage for it.

aeoranday commented on src/BinaryFileValidator.cpp:
Is there a reason you want this class to be header-only? I don't see why this cannot follow the header and source file structure.

aeoranday commented on src/BinaryFileValidator.cpp:
```diff
@@ -0,0 +1,78 @@
+            std::to_string(bytes_read) + ")";
+    return false;
+  }
+  if (!stream.good() && !stream.eof()) {
```
I'm a bit confused on this condition. Why do we need the condition on EOF?

aeoranday commented on include/tpglibs/testapp/common/BinaryFileValidator.hpp:
```diff
@@ -0,0 +1,98 @@
+class BinaryFileValidator {
+ public:
+  static constexpr uint32_t MAGIC_NUMBER = 0x54504754;  ///< "TPGT"
+  static constexpr uint32_t VERSION = 0x010004;         ///< 1.0.4 in hex
```
Where is this version derived from? Is this the `tpglibs` version at the time of this creation? We cannot keep this number up-to-date with the `tpglibs` version.

xinyue-uoft replied:
This is indeed the version at creation. The hope is to create a marker for backward compatibility. The original goal is that in the future if there is a breaking change to the format of the binary files, then we update this value here to mark the end of compatibility with previously generated files.

Maybe we can just keep this independent of `tpglibs` version?

aeoranday commented on include/tpglibs/testapp/common/BinaryFileValidator.hpp:
```diff
@@ -0,0 +1,98 @@
+ public:
+  static constexpr uint32_t MAGIC_NUMBER = 0x54504754;  ///< "TPGT"
+  static constexpr uint32_t VERSION = 0x010004;         ///< 1.0.4 in hex
+  static constexpr size_t HEADER_SIZE = sizeof(BinaryFileHeader);
```
For static member variables, we prefer the style `ClassName.s_variable_name`: https://dune-daq-sw.readthedocs.io/en/latest/packages/styleguide/#24-variable-names

aeoranday commented on src/BinaryFileValidator.cpp:
```diff
@@ -0,0 +1,78 @@
+  if (header.version != VERSION) {
+    error = "Unsupported binary file version";
+    return false;
+  }
```
It may be useful to also indicate the found magic number and version vs the expected magic number and version.

aeoranday commented on src/TestConfigParser.cpp:
```diff
@@ -0,0 +1,319 @@
+  return true;
+}
+
+bool TestConfigParser::parse_tpgenerator_config(const nlohmann::json& config,
```
This function seems to make use of features already defined by `validate_required_*`. Is there a reason you don't use those here? Should there be a separated `ConfigValidator` class?

xinyue-uoft replied:
Fix the leftover manual validation. There is no separate `ConfigValidator` class because these validation methods are only specific to configurations and is not expected to be used elsewhere.

aeoranday commented on src/TestConfigParser.cpp:
Is there a reason you want this class to be header-only? I don't see why this cannot follow the header and source file structure.

aeoranday commented on include/tpglibs/testapp/frame/FrameReader.hpp:
```diff
@@ -0,0 +1,167 @@
+/**
+ * @brief Status codes for frame reading operations
+ */
+enum class FrameReadStatus {
+  OK,           ///< Frame read successfully
+  END_OF_FILE,  ///< End of file reached (normal condition)
+  ERROR         ///< Fatal error (invalid header, malformed frame, I/O error)
+};
```
Our style prefers enum variables to follow `kVariableName`: https://dune-daq-sw.readthedocs.io/en/latest/packages/styleguide/#27-enumerator-names

I think `kEOF` would also be acceptable for how universal the EOF acronym is here.

aeoranday commented on include/tpglibs/testapp/tp/TPValidator.hpp:
```diff
@@ -0,0 +1,89 @@
+ * @brief Result of TP comparison
+ */
+struct TPComparisonResult {
+  bool matches;                    ///< True if TPs match exactly
+  size_t first_mismatch_index;     ///< Index of first mismatch (SIZE_MAX if matches)
+  std::string mismatch_reason;     ///< Description of mismatch
```
Same comment on style: `kVariableName`.

xinyue-uoft replied:
This is a struct instead of a enum, therefore, I think it should have k prefix

aeoranday commented on include/tpglibs/testapp/tp/TPComparator.hpp:
```diff
@@ -0,0 +1,89 @@
+   */
+  static TPComparisonResult compare(
+      const std::vector<dunedaq::trgdataformats::TriggerPrimitive>& expected,
+      const std::vector<dunedaq::trgdataformats::TriggerPrimitive>& actual);
```
The use of `expected` and `actual` makes this seem more like a `TPValidator` than a `TPComparator`. With similar language on `TPValidationResult`, it is much more clear to me the objective of this class and what to expect from the contents of the outcome. The words of "compare" and "comparator" are still true, but they are vague in comparison to "validate" and "validator".

More specific to `compare` and `compare_single_tp`, the intuition seems backwards to me. I would first assume `compare` would operate on 2 TPs and expect to see `compare_tp_vectors`. Given the class name `TPComparator`, I don't think it's necessary to be verbose for both functions, but that is also an option.

aeoranday replied:
I see later `compare_single_tp`. This seems backwards to me. I would think that `compare` would be single TPs and `compare_tp_vectors` would be comparing TP vectors. The other option is to have full verbosity for both.

aeoranday commented on src/TPComparator.cpp:
```diff
@@ -0,0 +1,146 @@
+        << " TPs, got " << actual_sorted.size() << " TPs";
+    result.mismatch_reason = oss.str();
+    return result;
+  }
```
This should be before sorting.

aeoranday commented on src/TPComparator.cpp:
```diff
@@ -0,0 +1,146 @@
+        return a.channel < b.channel;
+      }
+      // Tertiary: samples_over_threshold
+      return a.samples_over_threshold < b.samples_over_threshold;
```
You may want to make use of [std::tie](https://en.cppreference.com/w/cpp/utility/tuple/tie).

aeoranday commented on src/TPComparator.cpp:
```diff
@@ -0,0 +1,146 @@
+    });
+}
+
+std::string TPComparator::compare_single_tp(
```
It may be better to create one long string of all the pieces that mismatch instead of only returning the first part. It would likely give more information to the tester what would be wrong. Some of it will be redundant, but it could avoid repeated testing to find finer details that were missed on the first pass.

aeoranday commented on include/tpglibs/testapp/tp/TPReader.hpp:
```diff
@@ -0,0 +1,114 @@
+ * - File header: 12 bytes (magic, version, reserved)
+ * - For each frame: frame_index (4 bytes), num_tps (4 bytes), TP data
+ */
+class TPReader {
```
Since this specifically reads the validation binary files, I think this should be `TPValidationReader`. Especially with functions such as `get_validation_frames`.

aeoranday commented on include/tpglibs/testapp/tp/TPReader.hpp:
```diff
@@ -0,0 +1,114 @@
+ private:
+  static constexpr size_t FILE_HEADER_SIZE = BinaryFileValidator::HEADER_SIZE;
+  static constexpr size_t FRAME_INDEX_SIZE = sizeof(uint32_t);
+  static constexpr size_t TP_COUNT_SIZE = sizeof(uint32_t);
```
`s_variable_name`

aeoranday commented on include/tpglibs/testapp/frame/DummyFrameAdapter.hpp:
```diff
@@ -0,0 +1,130 @@
+  using word_t = uint64_t;
+  static constexpr int s_num_channels = 64;
+  static constexpr int s_time_samples_per_frame = 256;
+  static constexpr int s_bits_per_adc = 16;
```
This should be `s_bits_per_adc = 14` to match the data frames we expect to receive. This may argue the need to check in `TPGenerator` if the data frame is the supported 14 b format. A 16 b format would be easy to support, but it would need to be treated differently. Other sizes would require much more thinking.

aeoranday commented on src/TPReader.cpp:
```diff
@@ -0,0 +1,147 @@
+  const size_t tp_size = sizeof(dunedaq::trgdataformats::TriggerPrimitive);
+  
+  // Reasonable maximum: 1 million TPs per frame (prevents DoS and overflow)
+  constexpr uint32_t MAX_TPS_PER_FRAME = 1000000;
```
At 64 channels and 256 time samples per frame, the actual maximum is 8192 TPs per frame.

The max TP count happens when the signal alternates from below threshold to above threshold for each time sample.

aeoranday commented on src/TPValidationReader.cpp:
```diff
@@ -0,0 +1,147 @@
+    // Read TPs
+    std::vector<dunedaq::trgdataformats::TriggerPrimitive> tps;
+    try {
+      tps.resize(num_tps);
```
Why is this not `reserve(num_tps)`?

xinyue-uoft replied:
A few lines underneath, we do

`file.read(reinterpret_cast<char*>(tps.data()), expected_bytes);`

Therefore, we need a fully initialized data container at `tps.data()`.  `reserve()` doesn't do this.

aeoranday commented on test/apps/generate_example_tests.cxx:
```diff
@@ -0,0 +1,539 @@
+     return a.time_start < b.time_start;
+   if (a.channel != b.channel)
+     return a.channel < b.channel;
+   return a.samples_over_threshold < b.samples_over_threshold;
```
This can make use of `std::tie`.

aeoranday replied:
Given that you've defined this twice, it may be useful to define it once and have it used in both locations. This is what the `operator<` does in the `TriggerPrimitiveTypeAdapter`. I don't think it's a good idea to copy that from https://github.com/DUNE-DAQ/trigger, but you have also defined it with a slight difference.

aeoranday commented on test/apps/generate_example_tests.cxx:
```diff
@@ -0,0 +1,539 @@
+ 
+   out.write(reinterpret_cast<const char*>(&timestamp), sizeof(timestamp));
+   out.write(reinterpret_cast<const char*>(&another_key), sizeof(another_key));
+   out.write(reinterpret_cast<const char*>(adc_data.data()), adc_data.size() * sizeof(int16_t));
```
The `sizeof(int16_t)` should instead be the `s_bits_per_adc` from the data frame.

aeoranday commented on test/apps/generate_example_tests.cxx:
I do not follow the design choice for having the `create_spike_pattern`, `create_constant_pattern`, and `create_multi_channel_spikes` take `adc_data` as reference t and _not_ create and return something like `adc_data`.

xinyue-uoft replied:
This is just a temporary script for creating the simple tests. They are only here for reference at the moment and I intend to remove them from this PR at the end. 

I think we can have a separate PR + discussion on whether to include them in `tpglibs` and better design this generation scripts at that point.

aeoranday commented on test/apps/generate_golden_test_data.cxx:
This makes use of many of the same utility functions as `generate_example_test_data.cxx`. It would likely be useful to have them defined elsewhere.

xinyue-uoft replied:
Same as above, these are only temporary.

aeoranday commented on test/apps/generate_golden_test_data.cxx:
```diff
@@ -0,0 +1,333 @@
+ // Amplitude: threshold + 100
+ constexpr int16_t SPIKE_ADC_DELTA_ABOVE_THRESHOLD = 100;
+ 
+ // TPGenerator detects the TP when processing t = 106 in this setup
```
`TPGenerator` _generates_ the TP at `t = 106`. It detects it starting at `t = 100` in this case.

aeoranday commented on test/apps/generate_golden_test_data.cxx:
```diff
@@ -0,0 +1,333 @@
+ 
+ /**
+  * @brief Calculate absolute timestamp for TP time_start
+  * Formula matches TPGenerator: (t - samples_over_threshold) * sample_tick_difference + timestamp
```
Based on this description, this test verifies that the TP generated is equivalent to an expected TP.

However, the objective for creating these tests seems to want to validate that the generated TP is expected for the known signal. The TP's `time_start` should be `SPIKE_START_SAMPLE * SAMPLE_TICK_DIFFERENCE + frame_timestamp` since we want to know that we are accurately representing the injected signal that we control.

Since the `TPGenerator` (in the current state) is consistent between verification and validation, the comparison appears the same (at the moment), but we should really be aiming for the second case.

aeoranday commented on docs/testapp/BinaryFormatSpecification.md:
```diff
@@ -1,151 +1,273 @@
+┌─────────────────────┐
+│ Frame Header        │ (16 bytes)
+│ - timestamp (8)     │ uint64_t: Frame timestamp
+│ - another_key (8)   │ uint64_t: Additional header field
```
Why do we have this? I see later that you set it to an arbitrary value, so it seems like this is not necessary for our purposes. I'm aware the reference has it, but the value is relevant for them.

xinyue-uoft replied:
Indeed it is not really used here. It is purely carried over from the reference. Will be removed.

aeoranday commented on unittest/binary_file_validator_test.cxx:
```diff
@@ -0,0 +1,120 @@
+ */
+
+#define BOOST_TEST_MODULE binary_file_validator_test
+#define FMT_HEADER_ONLY
```
It doesn't look like you make use of the `fmt` library, so this is not needed. (This also goes for other unit tests.)

xinyue-uoft replied:
Indeed found the same thing in other unit tests. I removed `fmt` header from the tests that doesn't make use of it at all (kept the few where there are commented line, namely `unittest/avx_processors_test.cxx` and `unittest/avx_factory_test.cxx`

aeoranday commented on unittest/include/test_helpers.hpp:
I think this should be separated into a `unittest/include` directory.

aeoranday reviewed (commented):
I have not tested the functionality of this PR yet. This is only a result of reading with various style and design comments, questions, and suggestions.

One major comment is the header-only design of the test application files. Since none of the classes are general and require header-only, I would suggest that these get separated into the include­–source structure.

I'll test after these changes. One aspect I'm concerned about is the 16 b dynamic range in the dummy frame which does not match the expected 14 b signal. Since most of the applied data is composed of 0s, I would believe these "passed", but I believe high-valued testing would start to break.

aeoranday reviewed (changes requested):
(Missed this on previous review submission.)

xinyue-uoft committed:
convert testapp classes from header-only to separate source

xinyue-uoft committed:
Constrain dummy frame ADC values to 14-bit range [0, 16383] with clamping and warnings.

xinyue-uoft committed:
Update BinaryFormatSpecification documentation: clarifying the definition of `channel_t` and detailing the width of the TP channel value.

xinyue-uoft committed:
Fix eof condition.

xinyue-uoft committed:
naming convention update.

xinyue-uoft committed:
Error messages now show expected and found values for magic number and version. This improves debugging by displaying actual values encountered.

xinyue-uoft committed:
Refactored parse_tpgenerator_config to use validate_required_array helper.

xinyue-uoft committed:
Renamed FrameReadStatus enum values to use k prefix.

xinyue-uoft committed:
Renamed TPComparator to TPValidator and compare() to validate() for clarity. Renamed compare_single_tp() to compare() to fix counterintuitive naming.

xinyue-uoft committed:
Moved size mismatch check before sorting in TPValidator.

xinyue-uoft committed:
Replaced manual comparison logic with std::tie in TPValidator sort_tps.

xinyue-uoft committed:
TPValidator::compare now collects all mismatches instead of returning on first. This provides complete diagnostic information without repeated test runs.

xinyue-uoft committed:
Renamed TPReader to TPValidationReader.

alessandrothea commented on include/tpglibs/testapp/frame/DummyFrameAdapter.hpp:
```diff
@@ -0,0 +1,147 @@
+  static constexpr int16_t s_adc_min_value = 0;       // 0 - 14-bit min
+  
+  uint64_t timestamp;      ///< Frame timestamp (8 bytes)
+  uint64_t another_key;    ///< Additional header field (8 bytes)
```
`another_key` seems a bit random.  Something like`header` is more to the point

xinyue-uoft replied:
This was mentioned by Alejandro above. It is a carry over artifact from the original dummy frame definition https://github.com/DUNE-DAQ/datahandlinglibs/blob/develop/include/datahandlinglibs/ReadoutTypes.hpp that was referenced. Will be removed soon.

alessandrothea commented on CMakeLists.txt:
```diff
@@ -36,8 +36,21 @@ daq_add_unit_test(ProcessorInternalStateBufferManager_test LINK_LIBRARIES tpglib
+daq_add_application(tpglibs_generator_test test_tpg_generator_app.cxx TEST LINK_LIBRARIES tpglibs)
+
+# Add golden test data generator utility
+daq_add_application(generate_golden_test_data generate_golden_test_data.cxx TEST LINK_LIBRARIES tpglibs)
```
"Golden test data" is a little, how to say, mysterious.
What is golden? What test?
I would rather talk about `reference` + `data` + `patterns`

xinyue-uoft replied:
Yes. I am planning to remove these generate test scripts before merging this PR. It is better to include the tests in asset system directly.

alessandrothea commented on CMakeLists.txt:
```diff
@@ -36,8 +36,21 @@ daq_add_unit_test(ProcessorInternalStateBufferManager_test LINK_LIBRARIES tpglib
+daq_add_application(generate_golden_test_data generate_golden_test_data.cxx TEST LINK_LIBRARIES tpglibs)
+
+# Add example test files generator utility
+daq_add_application(generate_example_tests generate_example_tests.cxx TEST LINK_LIBRARIES tpglibs)
```
What does this app generate? tests or test data?

xinyue-uoft replied:
As explained above, these are temporary.

alessandrothea commented on include/tpglibs/testapp/frame/DummyFrameAdapter.hpp:
```diff
@@ -0,0 +1,147 @@
+namespace testapp {
+
+/**
+ * @brief Dummy frame structure for test applications
```
It seems to me that this is more than a "dummy" frame: this is a reference/minimal 14b dataframe.

xinyue-uoft replied:
Yes. The naming is more or less carried on from `datahandlinglibs` calling theirs `DUMMY_FRAME_STRUCT`.

alessandrothea reviewed (changes requested):
Great work.
I did not have the time to look into the code.
The comments about naming conventions: in general I would trade the use of `dummy` and `test` words for something more specific to the purpose of the application, which is the validation of the tpg chain by regression tests involving generation and comparison to reference patterns.

alessandrothea commented:
This is a summary of my comment at yesterday's Readout WG Meeting:

- inspection of the generated TPs and their analysis (potentially) is an important aspect for preparation of input patterns and debugging issues when detected by the tool
- my suggestion is to add, if not available already, a way to write output products (TPs) in a format easily consumable by analysis tools (e.g. python, pandas). Being able to inspect input waveforms would be a bonus.
- Depending on the typical output size, the "analysis" format options could include CSV or HDF5 or...

xinyue-uoft committed:
Rename static constexpr members in TPValidationReader to use s_ prefix.
This enforces DUNE DAQ style guide naming convention for static class data members.

xinyue-uoft committed:
Reduced maximum TP count validation limit from 1 million to 8192. This matches the actual physical constraint based on frame dimensions (64 channels × 256 time samples).

xinyue-uoft committed:
Remove unused another_key field from frame header, reducing it from 16 to 8 bytes.

xinyue-uoft committed:
Remove unused fmt library includes and defines from unit tests. The fmt library was included but never used in active code.

xinyue-uoft committed:
Move test_help.hpp to include directory.

xinyue-uoft committed:
Remove unused FMT_HEADER_ONLY defines from unit tests. The fmt library was not used in these files.

xinyue-uoft commented:
Addressed the all the review comments left over. 

Regarding the comments related to `generate_*_data.cxx`: Sorry for the confusion, but these files are not really part of this commit. I am including them here for context about the full data flow. The original plan is to directly have the generated configuration and binary files added to the asset management system.

aeoranday commented on test/apps/test_tpg_processor_app.cxx:
```suggestion
std::vector<int16_t>& expected = validation_data[validation_index];
expected.resize(16, 0);

```

There is a problem when using `samples_per_time_step < 16` as the `expected` vector would only be the size of `samples_per_time_step` while trying to complete the `mismatch` operation. There is a similar case that you had done for `time_step_data`.

aeoranday reviewed (changes requested):
I followed through the given test file generators and confirmed that the test application worked with those in most cases. The processor test application fails for `samples_per_time_step < 16`, and I've provided a suggestion on how to resolve it with details of the problem in the comment.

Otherwise, I followed through the documentation, made my own test file generator, and successfully tested those files with this test application suite.

xinyue-uoft committed:
Fix out-of-bounds read in validation when samples_per_time_step < 16.

Resize expected vector to 16 (zero-padded) to match the result array, consistent with how time_step_data is already handled.

Co-authored-by: Alejandro Oranday <alejandro@oran.day>

xinyue-uoft committed:
Fix expected vector initialization in validation step

Github did not remove duplicate initialization from last auto commit. Fixed now.

xinyue-uoft committed:
Remove const instantiation

aeoranday reviewed (approved):
LGTM. Thanks for making this!

xinyue-uoft committed:
Delete test/apps/generate_example_tests.cxx

xinyue-uoft committed:
Delete test/apps/generate_golden_test_data.cxx

xinyue-uoft committed:
Delete test/golden directory

xinyue-uoft committed:
Remove test data and example generator applications

Removed golden test data and example test files generator utilities from CMake configuration.

xinyue-uoft reviewed (commented):
Removed the example test generation and utilities that shouldn't live inside the test application PR

xinyue-uoft committed:
Dummy commit to rerun CI

alessandrothea reviewed (approved):
LGTM
